cgen error for `Name[int](123)`, when `type Name[T] = string | int | T`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.0 2fec2fa.c5e1c8e
OS: linux, Ubuntu 20.04.6 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /v/vnew
vexe: /v/vnew/v
vexe mtime: 2023-08-17 13:54:55

vroot: OK, value: /v/vnew
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.41.0
Git vroot status: weekly.2023.33-5-gc5e1c8ee
.git/config present: true

CC version: cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg a.v`
```v
type Name[T] = string | int | T
_ := Name[int](123)

```

**What did you expect to see?**

either a compiled program, or a checker error

**What did you see instead?**
```
==================
/tmp/v_1000/a.13613374991411628734.tmp.c:948: error: duplicate member '_int'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -447,6 +447,7 @@ fn (mut g Gen) gen_str_for_union_sum_type(info ast.SumType, styp string, typ_str
 		if typ.idx() in idxs {
 			continue
 		}
+		idxs << typ.idx()
 		typ_name := g.typ(typ)
 		mut func_name := g.get_str_fn(typ)
 		sym := g.table.sym(typ)
@@ -486,7 +487,6 @@ fn (mut g Gen) gen_str_for_union_sum_type(info ast.SumType, styp string, typ_str
 			}))'
 			fn_builder.write_string('\t\tcase ${int(typ)}: return ${res};\n')
 		}
-		idxs << typ.idx()
 	}
 	fn_builder.writeln('\t\tdefault: return _SLIT("unknown sum type value");')
 	fn_builder.writeln('\t}')

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -442,7 +442,11 @@ fn (mut g Gen) gen_str_for_union_sum_type(info ast.SumType, styp string, typ_str
 		clean_sum_type_v_type_name = util.strip_main_name(typ_str)
 	}
 	fn_builder.writeln('\tswitch(x._typ) {')
+	mut idxs := []int{}
 	for typ in info.variants {
+		if typ.idx() in idxs {
+			continue
+		}
 		typ_name := g.typ(typ)
 		mut func_name := g.get_str_fn(typ)
 		sym := g.table.sym(typ)
@@ -482,6 +486,7 @@ fn (mut g Gen) gen_str_for_union_sum_type(info ast.SumType, styp string, typ_str
 			}))'
 			fn_builder.write_string('\t\tcase ${int(typ)}: return ${res};\n')
 		}
+		idxs << typ.idx()
 	}
 	fn_builder.writeln('\t\tdefault: return _SLIT("unknown sum type value");')
 	fn_builder.writeln('\t}')

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -444,10 +444,10 @@ fn (mut g Gen) gen_str_for_union_sum_type(info ast.SumType, styp string, typ_str
 	fn_builder.writeln('\tswitch(x._typ) {')
 	mut idxs := []int{}
 	for typ in info.variants {
-		if typ.idx() in idxs {
+		if typ in idxs {
 			continue
 		}
-		idxs << typ.idx()
+		idxs << typ
 		typ_name := g.typ(typ)
 		mut func_name := g.get_str_fn(typ)
 		sym := g.table.sym(typ)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1000,9 +1000,14 @@ pub fn (mut g Gen) write_typeof_functions() {
 				tidx := g.table.find_type_idx(sym.name)
 				g.writeln('\tswitch(sidx) {')
 				g.writeln('\t\tcase ${tidx}: return "${util.strip_main_name(sym.name)}";')
+				mut idxs := []int{}
 				for v in sum_info.variants {
+					if v.idx() in idxs {
+						continue
+					}
 					subtype := g.table.sym(v)
 					g.writeln('\t\tcase ${int(v)}: return "${util.strip_main_name(subtype.name)}";')
+					idxs << v.idx()
 				}
 				g.writeln('\t\tdefault: return "unknown ${util.strip_main_name(sym.name)}";')
 				g.writeln('\t}')
@@ -1021,8 +1026,13 @@ pub fn (mut g Gen) write_typeof_functions() {
 				tidx := g.table.find_type_idx(sym.name)
 				g.writeln('\tswitch(sidx) {')
 				g.writeln('\t\tcase ${tidx}: return ${int(ityp)};')
+				mut idxs := []int{}
 				for v in sum_info.variants {
+					if v.idx() in idxs {
+						continue
+					}
 					g.writeln('\t\tcase ${int(v)}: return ${int(v)};')
+					idxs << v.idx()
 				}
 				g.writeln('\t\tdefault: return ${int(ityp)};')
 				g.writeln('\t}')
@@ -6678,12 +6688,21 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 				g.typedefs.writeln('typedef struct ${name} ${name};')
 				g.type_definitions.writeln('')
 				g.type_definitions.writeln('// Union sum type ${name} = ')
+				mut idxs := []int{}
 				for variant in sym.info.variants {
+					if variant.idx() in idxs {
+						continue
+					}
 					g.type_definitions.writeln('//          | ${variant:4d} = ${g.typ(variant.idx()):-20s}')
+					idxs << variant.idx()
 				}
+				idxs.clear()
 				g.type_definitions.writeln('struct ${name} {')
 				g.type_definitions.writeln('\tunion {')
 				for variant in sym.info.variants {
+					if variant.idx() in idxs {
+						continue
+					}
 					variant_sym := g.table.sym(variant)
 					mut var := if variant.has_flag(.option) { variant } else { variant.ref() }
 					variant_name := g.get_sumtype_variant_name(variant, variant_sym)
@@ -6694,6 +6713,7 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 					}
 					var_type := if variant.has_flag(.option) { '${g.typ(var)}*' } else { g.typ(var) }
 					g.type_definitions.writeln('\t\t${var_type} _${variant_name};')
+					idxs << variant.idx()
 				}
 				g.type_definitions.writeln('\t};')
 				g.type_definitions.writeln('\tint _typ;')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1002,12 +1002,12 @@ pub fn (mut g Gen) write_typeof_functions() {
 				g.writeln('\t\tcase ${tidx}: return "${util.strip_main_name(sym.name)}";')
 				mut idxs := []int{}
 				for v in sum_info.variants {
-					if v.idx() in idxs {
+					if v in idxs {
 						continue
 					}
 					subtype := g.table.sym(v)
 					g.writeln('\t\tcase ${int(v)}: return "${util.strip_main_name(subtype.name)}";')
-					idxs << v.idx()
+					idxs << v
 				}
 				g.writeln('\t\tdefault: return "unknown ${util.strip_main_name(sym.name)}";')
 				g.writeln('\t}')
@@ -1028,11 +1028,11 @@ pub fn (mut g Gen) write_typeof_functions() {
 				g.writeln('\t\tcase ${tidx}: return ${int(ityp)};')
 				mut idxs := []int{}
 				for v in sum_info.variants {
-					if v.idx() in idxs {
+					if v in idxs {
 						continue
 					}
 					g.writeln('\t\tcase ${int(v)}: return ${int(v)};')
-					idxs << v.idx()
+					idxs << v
 				}
 				g.writeln('\t\tdefault: return ${int(ityp)};')
 				g.writeln('\t}')
@@ -6690,19 +6690,20 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 				g.type_definitions.writeln('// Union sum type ${name} = ')
 				mut idxs := []int{}
 				for variant in sym.info.variants {
-					if variant.idx() in idxs {
+					if variant in idxs {
 						continue
 					}
 					g.type_definitions.writeln('//          | ${variant:4d} = ${g.typ(variant.idx()):-20s}')
-					idxs << variant.idx()
+					idxs << variant
 				}
 				idxs.clear()
 				g.type_definitions.writeln('struct ${name} {')
 				g.type_definitions.writeln('\tunion {')
 				for variant in sym.info.variants {
-					if variant.idx() in idxs {
+					if variant in idxs {
 						continue
 					}
+					idxs << variant
 					variant_sym := g.table.sym(variant)
 					mut var := if variant.has_flag(.option) { variant } else { variant.ref() }
 					variant_name := g.get_sumtype_variant_name(variant, variant_sym)
@@ -6713,7 +6714,6 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 					}
 					var_type := if variant.has_flag(.option) { '${g.typ(var)}*' } else { g.typ(var) }
 					g.type_definitions.writeln('\t\t${var_type} _${variant_name};')
-					idxs << variant.idx()
 				}
 				g.type_definitions.writeln('\t};')
 				g.type_definitions.writeln('\tint _typ;')

--- a/vlib/v/tests/generic_sumtype_repeated_test.v
+++ b/vlib/v/tests/generic_sumtype_repeated_test.v
@@ -1,0 +1,7 @@
+type Name[T] = T | int | string
+
+fn test_main() {
+	a := Name[int](123)
+	assert dump(a) == Name[int](123)
+	println(a)
+}

